app/commands/exercise/approach/link_matching_submissions.rb

```diff
@@ -0,0 +1,35 @@
+class Exercise::Approach::LinkMatchingSubmissions
+  include Mandate
+
+  initialize_with :approach
+
+  delegate :exercise, to: :approach
+
+  def call
+    return relink_submissions!(linked_submissions) if approach.tags.blank?
+
+    update_submissions!(linked_submissions)
+    update_submissions!(unlinked_submissions.tagged)
+  end
+
+  private
+  def linked_submissions = Submission.where(exercise:, approach:)
+  def unlinked_submissions = Submission.where(exercise:, approach: nil)
+
+  def update_submissions!(submissions)
+    submissions.find_each do |submission|
+      new_approach = approach.matching_tags?(submission.tags) ? approach : nil
+      submission.update(approach: new_approach)
```

Comment:
I'm not sure of this logic. I think what I'm seeing is that we find all solutions that are `nil` or this approach, and update them to either be `nil` or this approach. Is that correct? I think, if so, this could do with a comment, as I spent a while trying to work out the logic here. What confused me was whether we were overriding other approaches or not, as we **do** do that via the relink submissions.

Maybe adding this comment at the top of the file would help:

# This is intended to run when the tags on an approach changes.
# If the approach has no tags then we want to relink all solutions that are attached to it - this serves
# the case where we remove tags from an approach. In other cases, we want to find submission that
# either are untagged, or ones that were previously linked to this approach, and check whether they should
# be still linked to this approach. Note: this command will not reassign a submission that's linked to a different
# approach to the approach that's passed in. For that, use `Submission::LinkToMatchingApproach.defer`

Reply:
A further comment. I feel like if we're setting this to be `nil` here and previously it had an approach, we should instead run `Submission::LinkToMatchingApproach.defer(submission)` in case a different approach **would** match. I think that's the missing bit that's causing me confusion.

Reply:
Final comment - you pretty much always want to use `update!` not `update` - else we'll get silent failures.

Reply:
I've re-thought the logic a bit and refactored the code. Let me know if this version makes more sense